src/syntax/ast.rs

```diff
@@ -5,101 +5,108 @@ pub struct Resource {
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Comment {
-    pub body: String,
-}
-
-#[derive(Debug, PartialEq)]
-pub struct Section {
-    pub key: Key,
-    pub body: Vec<Entry>,
+pub struct Message {
+    pub id: Identifier,
+    pub value: Option<Pattern>,
+    pub attributes: Option<Vec<Attribute>>,
     pub comment: Option<Comment>,
 }
 
 #[derive(Debug, PartialEq)]
 pub enum Entry {
     Message(Message),
-    Comment(Comment),
     Section(Section),
+    Comment(Comment),
     Junk(JunkEntry),
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Message {
+pub struct Pattern {
+    pub elements: Vec<Expression>,
+    pub quoted: bool,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Expression {
+    Pattern(Pattern),
+    String(String),
+    Number(Number),
+    MessageReference(String),
+    ExternalArgument(String),
```

Comment:
I don't know how closely you'd like to keep the Rust implementation with the ASDL, so I'll let you decide here. The ASDL specifies the `id` field on `MessageReference` and `ExternalArgument`:

    MessageReference(iden id)
    ExternalArgument(iden id)

which in Rust would map to:

    MessageReference { id: Identifier },
    ExternalArgument { id: Identifier },

(which was in the ast.rs before this PR).